relay-server/src/metrics.rs

```diff
@@ -0,0 +1,200 @@
+    EventSizeBytesUncompressed,
+    /// Number of projects in the ProjectCache that are waiting for their state to be updated.
+    ProjectStatePending,
+    /// Number of project state requested from the Upstream for the current batch request.
+    ProjectStateRequestBatchSize,
+    /// Number of project states received from the Upstream for the current batch request.
+    ProjectStateReceived,
+    /// Number of project states currently held in the ProjectState cache.
+    ProjectStateCacheSize,
+}
+
+impl HistogramMetric for RelayHistograms {
+    fn name(&self) -> &'static str {
+        match self {
+            RelayHistograms::EventQueueSizePct => "event.queue_size.pct",
+            RelayHistograms::EventQueueSize => "event.queue_size",
+            RelayHistograms::EventSizeBytesRaw => "event.size_bytes.raw",
+            RelayHistograms::EventSizeBytesUncompressed => "event.size_bytes.uncompressed",
+            RelayHistograms::ProjectStatePending => "project_state.pending",
+            RelayHistograms::ProjectStateRequestBatchSize => "project_state.request.batch_size",
+            RelayHistograms::ProjectStateReceived => "project_state.received",
+            RelayHistograms::ProjectStateCacheSize => "project_cache.size",
+        }
+    }
+}
+
+/// Timer metrics used by Relay
+pub enum RelayTimers {
+    /// The time spent deserializing an event from a JSON byte array into a `relay_general::protocol::Event`.
+    EventProcessingDeserialize,
+    /// Time spent running event processors on an event.
+    /// Event processing happens before filtering.
+    EventProcessingProcess,
+    /// Time spent running filtering on an event.
+    EventProcessingFiltering,
+    /// Time spent checking for rate limits in Redis.
+    /// Note that not all events are checked against Redis. After an event is rate limited
+    /// for period A, any event using the same key coming during period A will be automatically
+    /// rate limited without checking against Redis.
+    EventProcessingRateLimiting,
+    /// Time spent in PII for the current event.
+    EventProcessingPii,
+    /// Time spent converting the event from an Annotated<Event> into a String containing the JSON
+    /// representation of the event.
+    EventProcessingSerialization,
+    /// Represents the time spent between receiving the event in Relay (i.e. beginning of the
+    /// request handling) up to the time before starting synchronous processing in the EventProcessor.
+    /// This is effectively the time between the event arriving in Relay and the event being ready
+    /// to be processed.
+    EventWaitTime,
+    /// This is the time the event spends in the EventProcessor (i.e. the sync processing of the
+    /// event).
+    /// It includes decoding the event envelope and the times spent in EventProcessingProcess,
+    /// `EventProcessingFiltering`, `EventProcessingRateLimiting`, `EventProcessingPii` and
+    /// EventProcessingSerialization.
```

Comment:
When referring to other metrics, let's use their string name. Users will not see our internal identifiers anywhere in statsd:

```rust
/// The time spent in synchronous event processing.
/// 
/// This timing covers the end-to-end processing in the CPU pool and comprises:
///  - `event_processing.deserialize`
///  - `event_processing.pii`
///  - `event_processing.serialization`
/// 
/// With Relay in processing mode, this includes the following additional timings:
/// 
///  - `event_processing.process`
///  - `event_processing.filtering`
///  - `event_processing.rate_limiting`
```

Also, note that a few metrics were missing